exp/orderbook/graph.go

```diff
@@ -478,3 +480,131 @@ func sortAndFilterPaths(
 
 	return filtered, nil
 }
+
+// makeTrade simulates execution of an exchange with a liquidity pool.
+//
+// It returns the amount that would be paid out by the pool for depositing
+// `amount` of `asset` alongside the new state of the liquidity pool after this
+// exchange, or an error. Errors can occur because of negative amounts, invalid
+// assets, invalid pools, XDR issues, etc.
+//
+// Refer to https://github.com/stellar/stellar-protocol/blob/master/core/cap-0038.md#pathpaymentstrictsendop-and-pathpaymentstrictreceiveop
+// for details on the exchange algorithm.
+func makeTrade(
+	asset xdr.Asset,
+	deposit uint32,
+	pool xdr.LiquidityPoolEntry,
+) (payout uint32, newPool xdr.LiquidityPoolEntry, err error) {
+	details, ok := pool.Body.GetConstantProduct()
+	if !ok {
+		err = errors.New("Liquidity pool unsupported: not constant product")
+		return
+	}
+
+	if !isAssetInLiquidityPool(asset, pool) {
+		err = errors.New("Can't exchange asset against liquidity pool")
+		return
+	}
+
+	depositXdr := xdr.Int64(deposit)
+	X, Y := details.ReserveA, details.ReserveB
+	exchangeReserve := details.ReserveB // amount of "other" asset in pool
+
+	// if necessary, swap the assets
+	isAssetA := details.Params.AssetA.Equals(asset)
+	if !isAssetA {
+		X, Y = details.ReserveB, details.ReserveA
+		exchangeReserve = details.ReserveA
+	}
+
+	payoutXdr, ok := calculatePoolPayout(X, Y, depositXdr, details.Params.Fee)
+
+	if !ok {
+		err = errors.New("Liquidity pool overflows from this exchange")
+		return
+	}
+
+	if payoutXdr > exchangeReserve {
+		err = errors.New("Not enough reserve for this exchange")
+		return
+	}
+
+	newPool, err = copyPoolState(pool)
+	if err != nil {
+		err = errors.Wrap(err, "Failed to duplicate LP state")
+		return
+	}
+
+	payout = uint32(payoutXdr)
```

Comment:
`calculatePoolPayout` returns `xdr.Int64` so likely this should be `uint64` or `int64` but not `uint32`.
```suggestion
	payout = uint64(payoutXdr)
```

Reply:
Definitely, I think I fixed the types now.